### PR TITLE
[Merged by Bors] - feat(StrongTopology): add lemmas about `IsVonNBounded`

### PR DESCRIPTION
**Pull request by urkud**
Also add `UniformOnFun.ofFun` here and there to avoid type defeq abuse.

---
- [x] depends on: #15215 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Algebra/Module/StrongTopology.lean [hunk 1]
```diff
@@ -103,22 +104,23 @@ that this has nice definitional properties. -/
...
+        (UniformOnFun.uniformSpace E F 𝔖) := by
   rw [instUniformSpace, UniformSpace.replaceTopology_eq]
 
 @[simp]
 theorem uniformity_toTopologicalSpace_eq [UniformSpace F] [UniformAddGroup F] (𝔖 : Set (Set E)) :
     (UniformConvergenceCLM.instUniformSpace σ F 𝔖).toTopologicalSpace =
-      UniformConvergenceCLM.instTopologicalSpace σ F 𝔖 :=
-  rfl
+      UniformConvergenceCLM.instTopologicalSpace σ F 𝔖 := by
+  with_reducible_and_instances rfl
```
Comment (urkud) [inline Mathlib/Topology/Algebra/Module/StrongTopology.lean:120]:
```suggestion
      UniformConvergenceCLM.instTopologicalSpace σ F 𝔖 :=
  rfl
```

File: Mathlib/Topology/Algebra/Module/StrongTopology.lean [hunk 2]
```diff
@@ -92,7 +93,7 @@ instance instTopologicalSpace [TopologicalSpace F] [TopologicalAddGroup F] (𝔖
     (DFunLike.coe : (UniformConvergenceCLM σ F 𝔖) → (E →ᵤ[𝔖] F))
 
 theorem topologicalSpace_eq [UniformSpace F] [UniformAddGroup F] (𝔖 : Set (Set E)) :
-    instTopologicalSpace σ F 𝔖 = TopologicalSpace.induced DFunLike.coe
+    instTopologicalSpace σ F 𝔖 = TopologicalSpace.induced (UniformOnFun.ofFun 𝔖 ∘ DFunLike.coe)
```
Comment (eric-wieser) [inline Mathlib/Topology/Algebra/Module/StrongTopology.lean:96]: Can you explain what's going on here in the PR description?

Comment (urkud) [reply Mathlib/Topology/Algebra/Module/StrongTopology.lean:96]: Done.

File: Mathlib/Topology/Algebra/Module/StrongTopology.lean [hunk 3]
```diff
@@ -185,6 +192,61 @@ theorem hasBasis_nhds_zero [TopologicalSpace F] [TopologicalAddGroup F]
...
+        𝓟 {f : UniformConvergenceCLM σ F 𝔖 | MapsTo f s (b i)} := by
+  letI : UniformSpace F := TopologicalAddGroup.toUniformSpace F
+  haveI : UniformAddGroup F := comm_topologicalAddGroup_is_uniform
+  rw [(embedding_coeFn σ F 𝔖).toInducing.nhds_eq_comap,
+    UniformOnFun.nhds_eq_of_basis _ _ h.uniformity_of_nhds_zero]
+  simp [MapsTo]
+
+theorem nhds_zero_eq [TopologicalSpace F] [TopologicalAddGroup F] (𝔖 : Set (Set E)) :
+    𝓝 (0 : UniformConvergenceCLM σ F 𝔖) =
+      ⨅ (s : Set E) (_ : s ∈ 𝔖) (t : Set F) (_ : t ∈ 𝓝 0),
```
Comment (eric-wieser) [inline Mathlib/Topology/Algebra/Module/StrongTopology.lean:208]: nit: maybe a little nicer as
```suggestion
      ⨅ s ∈ 𝔖, ⨅ t ∈ 𝓝 (0 : F),
```

File: Mathlib/Topology/Algebra/Module/StrongTopology.lean [hunk 4]
```diff
@@ -185,6 +192,61 @@ theorem hasBasis_nhds_zero [TopologicalSpace F] [TopologicalAddGroup F]
...
+        𝓟 {f : UniformConvergenceCLM σ F 𝔖 | MapsTo f s t} :=
+  nhds_zero_eq_of_basis _ _ _ (𝓝 0).basis_sets
+
+variable {F} in
+theorem gen_mem_nhds_zero [TopologicalSpace F] [TopologicalAddGroup F]
```
Comment (eric-wieser) [inline Mathlib/Topology/Algebra/Module/StrongTopology.lean:213]: What's the meaning of `gen` here? If we don't have precedence for this name elsewhere, and you can't think of a better name, can you add a docstring expanding it?

Comment (urkud) [reply Mathlib/Topology/Algebra/Module/StrongTopology.lean:213]: I guess, it's from something like "generator". I copied from [UniformOnFun.gen_mem_nhds](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergenceTopology.html#UniformOnFun.gen_mem_nhds)

Comment (urkud) [reply Mathlib/Topology/Algebra/Module/StrongTopology.lean:213]: I'm going to reformulate it using `Eventually` and rename it.

**Review by ADedecker — APPROVED** [review]
I got nerdsniped and spent some time thinking about another approach (proving the result for UniformOnFun and then pulling it back with some extra API) but I don't want to delay this for no reason so:

bors r+

Thanks !

**Comment by urkud** [comment]
Let's try again
bors merge

**Pull request closed without merging**